src/backend/libpq/pqcomm.c

```diff
@@ -2080,3 +2083,50 @@ pq_setkeepalivescount(int count, Port *port)
 
 	return STATUS_OK;
 }
+
+/*
+ * Check if the client is still connected.
+ */
+bool
+pq_check_connection(void)
+{
+#if defined(POLLRDHUP) || defined(__darwin__)
+	/*
+	 * POLLRDHUP is a Linux extension to poll(2) to detect sockets closed by the
+	 * other end. OSX is able to make the same detection via POSIX-compliant
+	 * POLLHUP option.
+	 * We don't have a portable way to do that without actually trying to read
+	 * or write data on other systems. We don't want to read because that would
+	 * be confused by pipelined queries and COPY data. Perhaps in future we'll
+	 * try to write a heartbeat message instead.
+	 */
+	struct pollfd pollfd;
+	int         rc;
+
+	pollfd.fd = MyProcPort->sock;
+#ifdef POLLRDHUP
```

Comment:
Fixed

Reply:
I agree, but don't like fixed version much, because of copypaste.
Will it be better to move polling logic to separate function with two parameters?
`static bool poll_fd(short init_events, short tgt_revents)`

Or add additional logic which can costs additional processing time, but still acceptable.
```
bool
pq_check_connection(void){
	short rdhup_ev;
#if defined(POLLRDHUP)
	rdhup_ev = POLLRDHUP;
#elif defined(__darwin__)
	rdhup_ev = 0;
#else
	return true;
#endif
	/*...*/
	pollfd.events = POLLOUT | POLLIN | rdhup_ev;
	/*...*/
	else if (rc == 1 && (pollfd.revents & (POLLHUP | rdhup_ev)))
	/*...*/
}
```
Or any variation of this.

Reply:
Rewritten, thanks